### ISSUE TITLE
[meta] Unexpected benefits and losses around Overloads and damage multipliers

### ISSUE THREAD

**Issue by Bloodmallet** [SimCMinMax/WoW-BugTracker#255]
## Summary
Overloads happen with a delay after the main cast. Overloads check for Buffs and Debuffs on their own. Therefore these checks happen after the main cast was cast successfully. To still allow Overloads to be affected like their primary casts, expiration delays were added to Buffs and Debuffs. Therefore a main cast, that would normally consume a buff/debuff, flags the buff/debuff for expiration instead of immediately consuming it. The actual consumption happens with the same delay as Overloads would happen. This time frame and different approaches on when the actual flagging/consumption happens (on cast success vs on hit) result in windows of opportunity (buffing more casts than intended) and loss (buffing no casts at all).

## Claim
```
Overloads check for buffs and debuffs on their own and are independent of their main cast
```
Use 'Exposed Elements' (talent), cast Lightning Bolt -> Earth Shock. If Lightning Bolt overloads, the Overload is buffed by Exposed Elements.

### Case A - unexpected benefit
```
Lightning Bolt -> Earth Shock
```
If Lightning Bolt overloads, the Overload is buffed. Exposed Elements is not consumed, and one can get more benefit than probably intended.

### Case B - unexpected loss
```
Earth Shock -> Lightning Bolt -> Earth Shock -> Lightning Bolt
```
First Lightning Bolt and its Overload are empowered. But due to the first Exposed Elements lingering for the Overload to happen, the second Earth Shock simply refreshes the still existing Exposed Elements on the target. The already scheduled expiration after the Overload still happens, consuming the fresh application. Result: Only the first Lightning Bolt and its Overload are buffed in this scenario instead of the expected both.

### Case C - unexpected benefit 
```
Lava Burst -> Lightning Bolt / Icefury / Elemental Blast -> Earth Shock / Frost Shock / Earthquake
```
Everything is buffed. Main casts, Overload, and the follow up instant spender. The first cast schedules the expiration. During that wait the second cast happens before the Overload could occure and therefore all three are buffed.

### Case D - unexpected loss 
```
Elemental Blast / Icefury -> Lava Surge empowered Lava Burst -> Lightning Bolt
```
Elemental Blast / Icefury is not buffed. Elemental Blast Overload / Icefury Overload is buffed. Master of the Elements is consumed on Elemental Blast / Icefury main cast hit (without buffing it, because damage is calculated at cast success already, when the buff wasn't present), resulting in the unexpect not-buffing of Lightning Bolt.

## High Voltage
All of the above about Lightning Bolts Overloads holds true for Lightning Bolts second overloads, too.

## How to fix 
Because I have no idea about the actual code base, I have two ideas on how to approach this problem. Either all Overloads become simple dmg copies of their main casts (like current tier bonues Earth Shock and Frost Shock), but I assume this would take a lot of time to do, and would counter some class fantasy (and it could wake the sleeping dragon Chain Lightning Overloads...). Or one could mitigate the bad effects of B and D by adding stacks to Exposed Elements to counter B, and changing the consumption of D to 'on cast success' instead of 'on hit'. This way A and C would still persist but at least that's in the favor of the players instead of secretly having negative synergies.

### Related Issues
Case A - https://github.com/SimCMinMax/WoW-BugTracker/issues/237
Case B - https://github.com/SimCMinMax/WoW-BugTracker/issues/238
Case C - https://github.com/SimCMinMax/WoW-BugTracker/issues/156
Case D - https://github.com/SimCMinMax/WoW-BugTracker/issues/164

**Comment by Bloodmallet** [comment]
To add to the Master of the Elements unexpected loss issue. Arthon from Earthshrine discord brought to my attention, that Case D breaks with Lightning Bolt too.

```
Lightning Bolt 1 -> Lava Surge empowered Lava Burst -> Lightning Bolt 2
```
In this case only the Overloads from Lightning Bolt 1 are empowered and Master of the Elements is consumed before Lightning Bolt 2 is cast.

**Comment by Bloodmallet** [comment]
The Lightning Bolt case indicates that the expiration of a Buff/Debuff is not exactly scheduled on the Buff/Debuff, but the expiration event itself checks for the Buff/Debuff and seems to be delayed to after potential Overloads. Resulting in the consumption of a Buff/Debuff that wasn't even present when the main cast was successfully cast.

**Comment by Bloodmallet** [comment]
Fixed in 8.0.1.26788
